src/main/java/org/truffleruby/core/thread/ThreadManager.java

```diff
@@ -470,23 +471,41 @@ public void shutdown() {
         }
 
         final FiberManager fiberManager = Layouts.THREAD.getFiberManager(rootThread);
+
+        if (Thread.currentThread() != Layouts.FIBER.getThread(fiberManager.getRootFiber())) {
+            throw new UnsupportedOperationException("ThreadManager.shutdown() must be called on the root Fiber of the main Thread");
+        }
+
         try {
             if (runningRubyThreads.size() > 1) {
                 killOtherThreads();
             }
         } finally {
-            fiberManager.shutdown();
             cleanup(rootThread);
         }
     }
 
+    /**
+     * Kill all Ruby threads, except the currently executing Thread. Waits that the killed threads
+     * have finished their cleanup and killed their fibers.
+     */
     @TruffleBoundary
     private void killOtherThreads() {
+        final Thread currentJavaThread = Thread.currentThread();
+        final List<CountDownLatch> threadsToWait = Collections.synchronizedList(new ArrayList<>());
+
         while (true) {
             try {
                 context.getSafepointManager().pauseAllThreadsAndExecute(null, false, (thread, currentNode) -> {
-                    if (thread != rootThread && Thread.currentThread() == Layouts.THREAD.getThread(thread)) {
-                        exit(context, thread, currentNode);
+                    if (Thread.currentThread() != currentJavaThread) {
```

Comment:
This is a bit confusing to read. Maybe `currentJavaThread` should be renamed to something like `initiatingJavaThread`?

Reply:
Good idea. Naming is hard with so many threads around.